DDF-3896 Add Intrigue upload editor

SmithJosh opened the pull request:
#### What does this PR do?
This PR adds an attribute editor to the Intrigue uploader that allows users to set attribute values that will be applied on upload. It's similar to the existing metacard edit capability, but users can set attributes and upload in one step now. Administrators can configure which attributes users can set, and can require that users set certain attributes before uploading.

If the editor is configured without any attributes, the uploader reverts to its existing behavior.

#### Who is reviewing it? 
(please choose AT LEAST two reviewers that need to approve the PR before it can get merged)
@alexaabrd @vinamartin @djblue 

#### Select relevant component teams: 
@codice/ui 
#### Ask 2 committers to review/merge the PR and tag them here. If you don't know who to ask, you can request reviews in https://groups.google.com/forum/#!forum/ddf-developers .
(please choose ONLY two committers from below, delete the rest)
@andrewkfiedler
@brendan-hofmann 
#### How should this be tested? (List steps with links to updated documentation)
Build and verify that:
* The editor is not shown if no attributes are configured (attributes are configured from the Catalog UI Search metatype)
* If attributes are configured, the editor is shown in both the Upload route and the list editor
* Required attributes are marked with an asterisk
* Upload is not permitted if any required field is unset (has only blank values)
* Upload is not permitted if any set field (non-blank value) has validation issues
* Warnings indicating required attributes display once a user tries to upload without completing the form
* The attribute reset button clears all fields and warnings
* Attributes are overridden as expected, both single- and multi-value attributes
* Editing a metacard is not affected by these changes

#### Any background context you want to provide?
#### What are the relevant tickets?
[DDF-3896](https://codice.atlassian.net/browse/DDF-3896)

#### Screenshots (if appropriate)
<img width="1433" alt="screen shot 2018-06-14 at 12 17 29 pm" src="https://user-images.githubusercontent.com/3878142/41436817-ad5e02c0-6fd7-11e8-8b7d-2b94f07e3d6f.png">
<img width="1075" alt="screen shot 2018-06-27 at 11 08 51 am" src="https://user-images.githubusercontent.com/3878142/41991643-fc7a31ee-79fa-11e8-96e4-8bdd6629689a.png">

#### Checklist:
- [x] Documentation Updated
- [x] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

brendan-hofmann commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/config/ConfigurationApplication.java:
```diff
@@ -212,6 +213,68 @@
+      }
+
+      String[] kvPair = entry.split("=", 2);
+      String attrName = kvPair[0];
```
❗️ I think you could still get exceptions if you don't check the values or length of `kvPair` before accessing them.

SmithJosh replied:
I didn't consider lists with null strings because that never happens with metatype configuration. I'll add a check though to be safe. 

With nonnull strings, though, the split method will always produce an array with at least one element. And I only access the second after checking the length, so this should be safe.

brendan-hofmann commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/config/ConfigurationApplication.java:
```diff
@@ -212,6 +213,68 @@
+  }
+
+  public void setRequiredAttributes(List<String> requiredAttributes) {
+    this.requiredAttributes = requiredAttributes;
```
❓ Should the list of required attributes be validated against the list of editable attributes?

SmithJosh replied:
The way it works right now, required attributes are ignored if they're not shown in the editor. I could change it though if validation is preferred.

brendan-hofmann commented on catalog/ui/catalog-ui-search/src/test/java/org/codice/ddf/catalog/ui/config/ConfigurationApplicationTest.java:
```diff
@@ -155,4 +159,22 @@ public void testsetAttributeAliases() throws Exception {
   }
+
+  @Test
+  public void testSetAttributeEnumMap() {
```
❗️ We should have more tests for the edge cases mentioned above.

brendan-hofmann commented on ui/packages/catalog-ui-search/src/main/webapp/component/ingest-details/ingest-details.view.js:
```diff
@@ -123,6 +140,18 @@ module.exports = Marionette.LayoutView.extend({
     startUpload: function() {
+        this.triggerMethod('ingestDetails:before:start');
+        for (var i = 0; i < properties.requiredAttributes.length; i++) {
+            var attrName = properties.requiredAttributes[i];
```
❓ Maybe we should gather the required attributes before we announce, so we don't spam the user with multiple errors?

SmithJosh replied:
Good idea. I'll change it

brendan-hofmann commented on ui/packages/catalog-ui-search/src/main/webapp/component/ingest-editor/ingest-editor.less:
```diff
@@ -0,0 +1,42 @@
+@{customElementNamespace}ingest-editor {
+  .customElement;
+  animation: new-ingest-editor .250s 1 ease-in-out;
+  padding: 10px;
```
❓ Doesn't pixel padding have scaling issues? I thought we preferred `em` units.

SmithJosh replied:
Changed to use em

alexaabrd commented on ui/packages/catalog-ui-search/src/main/webapp/component/ingest-editor/ingest-editor.view.js:
```diff
@@ -0,0 +1,70 @@
+        onBeforeShow: function() {
+            this.editorProperties.show(PropertyCollectionView.generateFilteredPropertyCollectionView(
+                properties.editorAttributes,
+                properties.requiredAttributes,
```
I dont think this second argument actually does anything, I think it should be [] because its supposed to be the metacards to extract the values from

alexaabrd commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/config/ConfigurationApplication.java:
```diff
@@ -212,6 +213,68 @@
+    }
+
+    setEditorAttributes(displayedAttributes);
+    setAttributeEnumMap(attributeValueEnums);
```
✏️ Totally not necessary, but to make this easier to follow I would just do `this.attributeEnumMap = attributeValueEnums`

and then I would get rid of all the `EditorAttributes` configs in this file and just use AttributeEnumMap.keySet where you need it

alexaabrd replied:
✏️ Totally not necessary, but to make this easier to follow I would just do this.attributeEnumMap = attributeValueEnums

and then I would get rid of all the EditorAttributes configs in this file and just use enums.keySet where you need it

alexaabrd commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/config/ConfigurationApplication.java:
```diff
@@ -212,6 +213,68 @@
+    Map<String, Set<String>> attributeValueEnums = new HashMap<>();
+
+    for (String entry : entries) {
+      if (entry.isEmpty()) {
```
❓ should this be `entry.trim().isEmpty()`

alexaabrd commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/config/ConfigurationApplication.java:
```diff
@@ -212,6 +213,68 @@
+      }
+
+      String[] kvPair = entry.split("=", 2);
+      String attrName = kvPair[0];
```
❓ Same here - should attrName be trimmed?

shaundmorris commented:
@SmithJosh this is cool! I would recommend a doc update for it as well since it is new capability in Intrigue. @ricklarsen I would think at least a release note on it. Thoughts?

alexaabrd commented on catalog/ui/catalog-ui-search/src/main/resources/OSGI-INF/metatype/org.codice.ddf.catalog.ui.config.xml:
```diff
@@ -384,6 +384,18 @@
             required="false"/>
+
+        <AD name="Upload Editor: Attribute Configuration" id="attributeEnumMap" required="false" type="String" cardinality="100"
+            default=""
```
❓ Do we want defaults here so that the ui portion isnt blank by default?

alexaabrd replied:
oh i missed that it goes away completely if nothing's configured. That makes sense - ignore my comment 😅

SmithJosh commented:
Thanks @shaundmorris. I'll update the docs

andrewkfiedler commented:
Talked with @SmithJosh about this a bit, there are a few updates we need to make to the list uploader as well to make it compatible with these changes.

SmithJosh commented:
Rebased, added required attribute highlighting, and refactored validation to make these changes compatible. I also attached another screenshot that shows the uploader rejecting an upload because validation failed

SmithJosh commented:
build now

vinamartin commented:
Manually verified with Chrome 👍 lgtm

alexaabrd commented:
verified this works with firefox and edge too 💯

SmithJosh commented:
build now

andrewkfiedler commented on ui/packages/catalog-ui-search/src/main/webapp/component/property/property.less:
```diff
@@ -81,6 +88,25 @@
     z-index: 0;
   }
+
+  .for-error {
```
❗️ This can be simplified to:
```
.for-error {
    color: contrast(@negative-color);
    display: none;
    line-height: @minimumLineSize;
    padding-left: @minimumSpacing;
    background: @negative-color;
    height: auto;
  }

  &.failed-required-check.show-required-warning {
    .for-error {
      display: block;
    }
  }
```

andrewkfiedler commented on ui/packages/catalog-ui-search/src/main/webapp/js/ApplicationSetup.js:
```diff
@@ -158,5 +158,5 @@ require([
 
     require('js/ApplicationStart');
-    require('codice-icons/icons/codice.font');
+    require('codice-icons/codice.font.config');
```
❓ Why the update?

SmithJosh replied:
At some point, this was breaking because it couldn't find `codice-icons/icons/codice.font`. Not sure why that was, but it's definitely there now. I'll change it back.

SmithJosh commented on ui/packages/catalog-ui-search/src/main/webapp/component/ingest-editor/ingest-editor.view.js:
```diff
@@ -0,0 +1,73 @@
+                [],
+            ));
+            this.editorProperties.currentView.$el.addClass("is-list");
+            this.$el.addClass('is-editing');
```
Removed it

SmithJosh commented:
build now

SmithJosh commented:
Build now
